meta-mender-tegra/meta-mender-tegra-jetpack5/classes/tegra-mender-setup.bbclass

```diff
@@ -35,7 +35,10 @@ PREFERRED_PROVIDER_u-boot-fw-utils = "u-boot-fw-utils-tegra"
 PREFERRED_PROVIDER_libubootenv:tegra = "${@'libubootenv-fake' if d.getVar('PREFERRED_PROVIDER_virtual/bootloader').startswith('cboot') else 'libubootenv'}"
 PREFERRED_RPROVIDER_u-boot-fw-utils = "u-boot-fw-utils-tegra"
 PREFERRED_RPROVIDER_libubootenv-bin:tegra = "${@'libubootenv-fake' if d.getVar('PREFERRED_PROVIDER_virtual/bootloader').startswith('cboot') else 'libubootenv-bin'}"
+PREFERRED_PROVIDER_virtual/bootloader:tegra194 = "edk2-firmware-tegra"
 PREFERRED_PROVIDER_libubootenv:tegra234 = "libubootenv-fake"
+PREFERRED_PROVIDER_libubootenv:tegra194 = "libubootenv-fake"
```

Comment:
As it is using libubootenv-fake, when the file /var/lib/mender/upgrade_available is deleted?

Mender is setting this "variable" inside ArtifactInstall
https://github.com/mendersoftware/mender/blob/4.0.0/support/modules/rootfs-image#L196
and the libubootenv-fake is creating the file here
https://github.com/OE4T/meta-mender-community/blob/5a1f806716a8c529a4b7a5ea7f25eebd3905952b/meta-mender-tegra/meta-mender-tegra-jetpack5/recipes-bsp/u-boot/libubootenv-fake/fw_setenv#L6
**Note: mender is using the command "${SETENV} -s" (with -s).**

The system reboot and boot from the new partition. In this scenario both Slots have "status: normal" when checking with "nvbootctrl dump-slots-info". mender-update-verifier.service wouldn't change anything because efivar "RootfsStatusSloatA" and "RootfsStatusSloatB" have size equal to zero.

Mender artifact commit happens here
https://github.com/mendersoftware/mender/blob/4.0.0/support/modules/rootfs-image#L238
and it uses the command "${SETENV} upgrade_available 0" (**without -s**), that means, this command is ignored by libubootenv-fake due the first if
https://github.com/OE4T/meta-mender-community/blob/5a1f806716a8c529a4b7a5ea7f25eebd3905952b/meta-mender-tegra/meta-mender-tegra-jetpack5/recipes-bsp/u-boot/libubootenv-fake/fw_setenv#L2

Could fw_setenv be something like?

```
#!/bin/sh

create_or_remove_file()
{
    if [ "$1" = "1" ]; then
        touch /var/lib/mender/upgrade_available
    else
        rm -f /var/lib/mender/upgrade_available
    fi
}

# To handle calls like "fw_setenv upgrade_available 0"
if [ "$1" != "-s" ] && [ "$#" -eq 2 ]; then
    if [ "$1" != "upgrade_available" ]; then
      exit 0
    fi

    create_or_remove_file $2
    exit 0
fi

# To handle calls like "fw_setenv -s <<EOF \
#mender_boot_part=$passive_num \
#mender_boot_part_hex=$passive_num_hex \
#upgrade_available=0 \
#EOF"
while IFS="= " read var restofline; do
    if [ "$var" != "upgrade_available" ]; then
        continue
    fi

	create_or_remove_file $restofline
done
```

Reply:
So you are saying that what happens in this file is not enough?
https://github.com/OE4T/meta-mender-community/blob/kirkstone/meta-mender-tegra/meta-mender-tegra-jetpack5/recipes-mender/mender-update-verifier/files/mender-update-verifier.sh

This seems to have worked reliably

Reply:
Let me rephrase my questions...

1. Are you running Mender client in a standalone or managed mode?
2. After updating the board with a `.mender` artifact, does the file `/var/lib/mender/upgrade_available` exist in your system?

Reply:
1. Mender client is ran in managed mode
2. The file `/var/lib/mender/upgrade_available` does not exists in my system after an update

Reply:
FYI, I tested your branch and all worked fine, the file is being removed as expected.
Something is wrong with my previous implementation or it is related to Mender 4.0.

Reply:
@djusten @mwest90  I have some doubts about how `upgrade_available` works. Please, correct me if I'm wrong

1. During ArtifactInstall this file (flag) is created when `mender-client` calls `fw_setenv -s ...`
2. Just after reboot and before `mender-client` starts, due to service dependency `mender-update-verifier` is run
3. It checks if upgrade has completed successfully and in that case removes `upgrade_available`
4. When `mender-client` starts, if `upgrade_available` still exists (upgrade_available=1) then upgrade has failed and `mender-client` should report failure (and rollback?) . On the other hand, if if `upgrade_available`  is missing (upgrade_available=0) then `mender-client` should report success.

Is that right?

Reply:
@irodriguez-veridas is this conversation resolved?

Reply:
Not really, I have not received any reply. @djusten or @mwest90 can you confirm the process is as I have described it? I just want to understand it, I have tested valid update artifacts can be successfully deployed and invalid ones (adding a wrong capsule file, for example) are rejected.

Reply:
My sincere apologize @irodriguez-veridas , I simply lost that you have asked me few question long time ago.
The problem I mentioned in the first comment was because I'm using Mender 4.0. `fw_setenv` has that  ["-s" protection](https://github.com/OE4T/meta-mender-community/blob/kirkstone/meta-mender-tegra/meta-mender-tegra-jetpack5/recipes-bsp/u-boot/libubootenv-fake/fw_setenv#L2), but Mender 4.0.x doesn't use the "-s" for setting the upgrade_available to 0 [here](https://github.com/mendersoftware/mender/blob/master/support/modules/rootfs-image#L259C19-L259C36). To solve my specific issue with Mender 4.0, I changed Mender to use ```${SETENV} -s - <<EOF
upgrade_available=0
EOF``` instead.
Please keep in mind that this does not apply for Mender 3.5.x.

---
Answering your question...

> 1.During ArtifactInstall this file (flag) is created when mender-client calls fw_setenv -s ... 

Correct.

> 2. Just after reboot and before mender-client starts, due to service dependency mender-update-verifier is run

Correct.

>It checks if upgrade has completed successfully and in that case removes upgrade_available

No, it checks if any rootfs slot has the status "unbootable", in that case it runs `fw_setenv -s <<< upgrade_available=0` and remove the `upgrade_available` flag/file.

>When mender-client starts, if upgrade_available still exists (upgrade_available=1) then upgrade has failed and mender-client should report failure (and rollback?) . On the other hand, if if upgrade_available is missing (upgrade_available=0) then mender-client should report success.

If the system/user try to do a mender commit and `upgrade_available=0` (file does not exist) it will say that mender install failed and should rollback. Logs ( I'm using mender 4.0.x):
```
$ mender-update commit 
...
record_id=4 severity=info time="2024-Jul-25 10:53:23.703428" name="Global" msg="Update Module output (stderr): Unexpected `upgrade_available=0` in ArtifactCommit." 
record_id=5 severity=info time="2024-Jul-25 10:53:23.703748" name="Global" msg="Update Module output (stderr): Upgrade failed and was reverted: refusing to commit!" 
record_id=6 severity=error time="2024-Jul-25 10:53:23.709678" name="Global" msg="Commit failed: Process returned non-zero exit status: ArtifactCommit: Process exited with status 1" 
record_id=7 severity=info time="2024-Jul-25 10:53:23.720356" name="Global" msg="Running State Script: /var/lib/mender/scripts/ArtifactCommit_Error_00" 
...
```

Reply:
Great! Thank you very much for the clarification @djusten !